### PR TITLE
client API: fix missing property change events after property updates

### LINKED ISSUE #4195: libmpv: playback-time not updated on a frame basis when observing with mpv_observe_property
### mpv version and platform

MacOS Sierra 10.12.3
mpv: HEAD-9740383, 0.24, 0.23, 0.22
ffmpeg: 3.2.4

### Reproduction steps
I'm using libmpv with opengl-cb in my opengl application. Like in a video player I want to know the current position or current frame in a playing video.
I used mpv_observe_property for that. With recent mpv versions this is not working the property is only updated once in a while (also the properties 'percent-pos' and 'time-pos')

If I remember it correctly it worked with older mpv versions.

To reproduce:
Use the sdl example from https://github.com/mpv-player/mpv-examples.git

Observe the property:
```
mpv_observe_property(mpv, 0, "playback-time", MPV_FORMAT_DOUBLE);
```

and wait for property changed events in the event loop, like this:

```
if (mp_event->event_id == MPV_EVENT_PROPERTY_CHANGE) {
    mpv_event_property *prop = (mpv_event_property *)mp_event->data;
    printf("MPV_EVENT_PROPERTY_CHANGE %s\n", prop->name);
    if (strcmp(prop->name, "playback-time") == 0 && prop->format == MPV_FORMAT_DOUBLE) {
        double pos = *(double*)prop->data;
        printf("playback-time: %f\n", pos);
    }
}

```
### Expected behavior

Should update the playback-time every frame.

### Actual behavior

Updates the property only sometimes. See attached log.
A test with mpv 0.18 showed it updates the property every second frame (see attached log)

### Log file

with mpv master: http://sprunge.us/XgQg
with mpv 0.18: http://sprunge.us/gXUT

### Sample files

Any video.

### PR DESCRIPTION
**Pull request by gunnarmarten**
When update_prop() successfully fetched a changed property value, it
set prop->changed to true to indicate the success.
mark_property_changed() also always set
prop->changed to true and additionally prop->need_new_value to true
This is the case since 6ac0ef78
If the observed property changes every frame and then due to timing
the next mark_property_changed() is called before
gen_property_change_event() and therefore directly after update_prop(),
prop->need_new_value was again true and indicated that a new value
has to be retrieved with update_prop(). As a result the event for the
last successful update_prop() was never triggered. This meant that
a property change event was never generated for frame-based properties
only for properties that were observed with MPV_FORMAT_NONE or when the
timing was different and gen_property_change_event() was called after
update_prop().
    
To fix this, mark_property_change() and update_prop() should not use the
same flag to indicate different things and therefore a new flag for
successfully update a property is introduced.

Fixes #4195

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

**Comment by jeeb** [comment]
Thanks for the contribution. Since the mentioned commit is by @rossy , highlighting him so he might take a look at it as well.

File: player/client.c [hunk 1]
```diff
@@ -1591,35 +1594,40 @@ static bool gen_property_change_event(struct mpv_handle *ctx)
     ctx->lowest_changed = ctx->num_properties;
     for (int n = start; n < ctx->num_properties; n++) {
         struct observe_property *prop = ctx->properties[n];
-        if ((prop->changed || prop->updating) && n < ctx->lowest_changed)
+        if ((prop->changed || prop->updating || prop->updated) && n < ctx->lowest_changed)
```
Comment (wnoun) [inline player/client.c]: This is to check properties whose value is being retrieved, so `(prop->changed || prop->updating || prop->updated)` can be simplified to `(prop->need_new_value || prop->updating)`, as event can be generated immediately in other cases.

**Comment by wnoun** [comment]
it works

File: player/client.c [hunk 2]
```diff
@@ -1591,35 +1594,40 @@ static bool gen_property_change_event(struct mpv_handle *ctx)
...
             ctx->lowest_changed = n;
         if (prop->changed) {
             bool get_value = prop->need_new_value;
             prop->need_new_value = false;
             prop->changed = false;
-            if (prop->format && get_value) {
+            if (get_value) {
```
Comment (wnoun) [inline player/client.c]: You shouldn't remove the check for `prop->format`. [the initial value of `prop->need_new_value` and `prop->changed` are both set to `true`](https://github.com/mpv-player/mpv/blob/4d001bb30db6786a44658b0625dc31027767a393/player/client.c#L1466). it tries giving user a notification immediately after user observes a property. so if user observes a property with `MPV_FORMAT_NONE`([osc.lua does this](https://github.com/mpv-player/mpv/blob/4d001bb30db6786a44658b0625dc31027767a393/player/lua/osc.lua#L2284)), the `update_prop` will try retrieving a value with `MPV_FORMAT_NONE`, which causes segment fault. Changing the initial value of `need_new_value` to `format != MPV_FORMAT_NONE` should also fix it.

Comment (gunnarmarten) [reply player/client.c]: Thanks for the review.
Good Catch! My bad, I somehow missed that.

I will take another look and fix it.

Comment (gunnarmarten) [reply player/client.c]: just pushed a fix 1d5b202.
As a bonus we can drop the need_new_value flag completely which makes the code a bit more readable.

**Comment by gunnarmarten** [comment]
I added a fix and resolved the conversation. @wnoun can you give it another spin?

**Comment by wnoun** [comment]
the issue is gone

**Comment by gunnarmarten** [comment]
Ping @jeeb or @rossy. Hoping to get this merged. Can you take a look?

I just rebased against master. The 2 commits should be squashed before merging. I can do that if you prefer.

**Comment by mseeley** [comment]
Kind greetings @wnoun, @jeeb, @rossy 👋.

We're also noticing similar behavior as @gunnarmarten; `playback-time` is eventing approximately once every 5s using libmpv on macOS Mojave in 0.28. Do you all think this PR is in a landable state?

**Pull request merged**